elastic/elasticsearch
# Add default ILM policies and switch to ILM for apm-data plugin

## What was done for 8.15?

The APM server team migrated management of index templates, component templates, ingest pipelines, etc from the [APM integration](https://github.com/elastic/integrations/tree/main/packages/apm) to the [APM data plugin](https://github.com/elastic/elasticsearch/tree/main/x-pack/plugin/apm-data).

As part of the migration, the default data retention (shipped with the integration) migrated from [Index Lifecycle Management](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-lifecycle-management.html) (ILM) to [Datastream Lifecycle](https://www.elastic.co/guide/en/elasticsearch/reference/current/data-stream-lifecycle.html) (DSL).

### Estimated Impact

We estimated NO impact on our users, for the most part. This is because APM, by default, ships all its data streams with default retention policies and does not make use of tiered storage by default. These default policies are *Managed* and thus assumed not to be tampered with by the users. To facilitate the customization of retention policies, the APM index templates have defined *custom* component templates that provide the mechanism for configuring a new ILM policy to be used instead of the default one. We did not change customer facing documentation, still advising users to leveraging ILM where necessary for customization.

### Actual Impact

Actually, the switch to the ES apm plugin caused several issues for our customers. The two main problems are listed as [known issues in the APM Server docs](https://www.elastic.co/guide/en/observability/8.15/apm-known-issues.html). 

Feedback from the Field   
*We have received feedback from enterprise customers that ILM provides flexible control over data management that DSL lacks, limiting enterprises’ ability to adjust data placement dynamically. For larger, centralized customers, ILM is essential as it enables frequent tier adjustments to avoid overloading costly Hot storage. While DSL may suit out-of-the-box solutions for certain markets, it does not meet the flexibility and control demands of many enterprise customers.*

## Path forward for 8.16

***We want to roll back to ILM from DSL, at least until 9.0***

Based on feedback from and several discussions with the field, we have agreed to roll back to ILM at least until 9.0. This is considered to be in the best interest of APM customers for the time being.   
The main arguments for switching back to ILM are:

* With the current solution, APM defaults to DSL. However many users require customization, for which we still recommended them to leverage ILM. Being compatible with ILM was a requirement to introduce DSL in the first place, as otherwise it would have been a breaking change during 8.x.   
* ILM is more suited for enterprise customers on stateful as it supports tiered storage configuration which is not supported by DSL.   
* If leveraging DSL for customization, users would need to use the PUT API, which is not making customization easier for them.   
* With the current state, we lost consistency. When creating a new 8.15 deployment, the default is DSL. When upgrading from older versions without customization, the default is DSL. When upgrading from older versions with customization, the default on 8.15 is ILM. The inconsistency is not only between deployments, but also within deployments if some datastreams are customized (ILM) where others are default (DSL). 


Overall we acknowledge that the change led to more complexity and confusion instead of making data management more straightforward and easier for our customers. 

We should consider whether DSL would be the preferred path forward for 9.0, but revert the changes for 8.x   

## Local Testing

### Tests are done with the following PRs/commits:

* [https://github.com/elastic/elasticsearch/pull/116240](https://github.com/elastic/elasticsearch/pull/116240)   
* [https://github.com/elastic/elasticsearch/pull/115687](https://github.com/elastic/elasticsearch/pull/115687) \- cherry-picked onto 8.16 branch

### Testing steps for local testing:

1. Build docker image locally using ./gradlew buildAarch64DockerImage  
2. Use the appropriate image to run the full stack with persistence enabled  
3. Install APM integration if stack version \< 8.15.0  
4. Send some data  
5. Assert that datastreams are created (also check for hidden datastreams)  
6. Assert that datastreams are managed correctly, example: GET /\_data\_stream/traces-apm-default  
   1. For \<8.15.0 version or clusters using APM and upgraded to 8.15.x should have their APM datastreams managed by ILM  
   2. For clusters created in 8.15.x their APM datastreams should be managed by DSL  
   3. On upgrade to 8.16.0 all APM datastreams should be managed by ILM  
7. Ensure there are no relevant errors in Elasticsearch or APM Server logs

<details>
<summary>Testing results</summary>

### New 8.16.0 cluster

* All the default ILM policies are installed in the cluster as expected  
* APM datastreams are managed by ILM policy for both normal and hidden datastreams

|  |  |
| ------------- | ------------- |
| <img width="634" alt="img1" src="https://github.com/user-attachments/assets/e96e86e3-86e4-43ce-b699-8978dfaaaaf3">  | <img width="833" alt="img2" src="https://github.com/user-attachments/assets/f54f9189-f1cc-4b84-b6b5-9e179401d0b9">  |

* Ingestion works as expected

### Upgrade from 8.12.2 to 8.16.0

* 8.12.2 cluster is managed by ILM for both normal and hidden datastreams

|  |  |
| ------------- | ------------- |
| <img width="674" alt="img3" src="https://github.com/user-attachments/assets/776cf018-93e2-47d1-b812-98fc344186cf"> | <img width="791" alt="img4" src="https://github.com/user-attachments/assets/473ae898-3a09-41b5-b46a-43075b9bb951"> |

* On upgrade to 8.16.0 and without sending data, the rollover on write flag was true for both normal and hidden datastreams \- indicating that lazy rollover was invoked for the datastreams  
* The indices continued to be managed by ILM after upgrade to 8.16.0 as expected

|  |  |
| ------------- | ------------- |
| <img width="650" alt="img5" src="https://github.com/user-attachments/assets/168d14bc-6bc1-4b2d-9437-74eb7306da10"> | <img width="792" alt="img6" src="https://github.com/user-attachments/assets/83963194-f146-4535-ab09-75a7f7f178ad"> |

* No relevant error in Elasticsearch logs was observed (there was an unrelated error for failing to download geoip database  
* Overriding with custom ILM template by using @custom component template works as expected

<img width="729" alt="img7" src="https://github.com/user-attachments/assets/fed7b500-9cb8-46f3-85f3-fc3957fc37ae">

* Ingestion worked without any errors post-upgrade

### Upgrade from a new 8.15.3 to 8.16.0

* 8.15.3 cluster is managed by DSL for both normal and hidden datastreams

|  |  |
| ------------- | ------------- |
| <img width="647" alt="img8" src="https://github.com/user-attachments/assets/82b33b73-ef2e-40f5-9bd7-d04d98ab7b99"> | <img width="783" alt="img9" src="https://github.com/user-attachments/assets/dff15e99-bf26-4c65-9944-2b9dd446a1a0"> |

* On upgrade to 8.16.0 and without sending any data the rollover on write flag was true for both normal and hidden datastreams \- indicating that lazy rollover was invoked for the datastreams  
* We can also see that on upgrade to 8.16.0 we migrate back to ILM (check the \`next\_generation\_managed\_by\` flag)
* Existing backing indices will be managed by DSL and any new indices will be managed by ILM.

|  |  |
| ------------- | ------------- |
| <img width="634" alt="img10" src="https://github.com/user-attachments/assets/caceb477-a68e-490c-89e0-3ec73005106c"> | <img width="803" alt="img11" src="https://github.com/user-attachments/assets/1a3243f0-c9c4-487c-82f5-b641c21b76b7"> |
</details>
